### ISSUE TITLE
Add scipy entropy functions

### ISSUE THREAD

**Issue by Uroc327** [jax-ml/jax#29596]
It would be very helpful to have the special distribution entropy functions implemented. E.g., `scipy.stats.poisson.entropy` is not trivial.

**Comment by michaeleliot** [comment]
Hey @Uroc327 , I took a stab at this, but haven't gotten it to work.

@jakevdp  I’d be grateful for any thoughts you might have on a better direction here, or on how my implementation could better align with the SciPy version. 

For context, SciPy uses a chunked summation estimate over the discrete probability mass function. You can see from the [discrete distribution _entropy function](https://github.com/scipy/scipy/blob/d46b9b31d3fb71b1e5fcd9bf9fa1e3a0b235c951/scipy/stats/_distn_infrastructure.py#L3846-L3852) as well as the [summation function](https://github.com/scipy/scipy/blob/d46b9b31d3fb71b1e5fcd9bf9fa1e3a0b235c951/scipy/stats/_distn_infrastructure.py#L3954-L3994) how this is done. [The poisson distribution](https://github.com/scipy/scipy/blob/b1296b9b4393e251511fe8fdd3e58c22a1124899/scipy/stats/_discrete_distns.py#L961-L1024) defaults to this implementation.

I have two problems with this:
1. The SciPy implementation introduces several magic values for chunk size, tolerance, and maximum allowed chunks. From my understanding, JAX generally prefers to avoid these. While it’s possible to specifically estimate discrete entropy for the poisson distribution via a power series of the [incomplete gamma integral](https://github.com/jeremybarnes/cephes/blob/60f27df395b8322c2da22c83751a2366b82d50d1/cprob/igam.c) (as Cephes does [here](https://github.com/jeremybarnes/cephes/blob/60f27df395b8322c2da22c83751a2366b82d50d1/cprob/pdtr.c#L3)), that approach introduces similar magic-number issues and is custom only for the poisson distribution. A rational function approximation could be worth exploring as an alternative if there is enough interest.
2. This kind of iterative summation with multiple breaking conditions seems difficult in jax. I've tried my best to rewrite it in a more jax friendly way, but haven't gotten the values to be close enough. I'm still fairly novice at this, so might be missing something obvious.


Below is my attempt. You can alternatively view [the full code on my branch](https://github.com/michaeleliot/jax/blob/232daef0e7af7e05f1976d72308e9dd70240ae53/jax/_src/scipy/stats/poisson.py#L121), as well as the [failing test](https://github.com/michaeleliot/jax/blob/232daef0e7af7e05f1976d72308e9dd70240ae53/tests/scipy_stats_test.py#L175-L189).  


```python
# /jax/_src/scipy/stats/poisson.py
from jax._src.scipy.special import entr #Only including additional imports
from jax._src.util import safe_map as map

# ... Other functions here ...

def entropy(mu: ArrayLike, loc: ArrayLike = 0) -> Array:
    mu, loc = promote_args_inexact("poisson.entropy", mu, loc)

    mu, loc = jnp.broadcast_arrays(mu, loc)
    mu_flat = jnp.ravel(mu)
    loc_flat = jnp.ravel(loc)
    median = mu_flat + 1/3 - 0.02 / mu_flat
    lb = jnp.asarray(0, dtype=np.float32)
    ub = jnp.asarray(np.inf, dtype=np.float32)

    def body_fn(args):
        mu_i, loc_i, median = args
        return _expect(mu_i, loc_i, lb, ub, median)
    combined = jnp.stack([mu_flat, loc_flat, median], axis=-1)
    result = map(body_fn, combined)

    flat = jnp.stack(result)
    result = jnp.reshape(flat, mu.shape)
    return result

def _expect(mu, loc, lb, ub, x0, inc = 1.0, maxcount=1000.0, chunksize=32.0):
    """JAX version of expectation helper with convergence and maxcount control."""
    x0 = jnp.clip(x0, lb, ub)

    tot_f = _sum_over_chunks(mu, loc, x0, ub + 1, inc, chunksize, maxcount)

    tot_b = _sum_over_chunks(mu, loc, lb, x0, inc, chunksize, maxcount)

    return tot_f + tot_b

def _sum_over_chunks(mu, loc, start, stop, inc, chunksize, max_chunks):
    """
    Split the range [start, stop) into chunks and sum over each chunk
    by calling _sum_over_range.
    """
    chunk_starts = start + chunksize * jnp.arange(max_chunks)
    chunk_stops = jnp.minimum(chunk_starts + chunksize, stop)

    # Function to sum a single chunk
    def sum_chunk(args):
        lb, ub = args
        return _sum_over_range(mu, loc, lb, ub, inc)

    # Map over all chunks
    combined = jnp.stack([chunk_starts, chunk_stops], axis=-1)
    result = map(sum_chunk, combined)
    flat = jnp.stack(result)
    return jnp.sum(flat)

def _sum_over_range(mu, loc, start, stop, inc):
    fun = lambda x: entr(pmf(x, mu, loc))
    xs = jnp.arange(start, stop, inc)
    vals = fun(xs)
    return jnp.sum(vals)
```

**Comment by jakevdp** [comment]
@michaeleliot - thanks for looking at this. That looks like a good start!

You're right that loopy implementations with value-dependent break statements are not great in JAX, though we do have a few special functions that are implemented similarly.

It would probably be a bit better to use `jax.lax.map` rather than Python's built-in `map` (or JAX's internal `safe_map`), because the latter rely on Python control flow and could generate very long programs. Also, `lax.map` has chunked batching built-in, so you may be able to take advantage of that to eliminate some of your boilerplate here.

**Comment by michaeleliot** [comment]
Hey @jakevdp , thanks for the response on this! I'll look into swapping out the map in that code. Good to know this looks like an ok direction.

**Comment by michaeleliot** [comment]
Took a second look. Still not much closer on why the estimates are off. What's surprising is that even a brute force estimate is still too low, which doesn't make any sense. I think I will leave this if someone else wants to take a look:

```python
def entropy(mu: ArrayLike, loc: ArrayLike = 0) -> Array:
 #Simplified version that still is too low for SciPy
  
  mu, loc = promote_args_inexact("poisson.entropy", mu, loc)

  mu, loc = jnp.broadcast_arrays(mu, loc)
  mu_flat = jnp.ravel(mu)
  loc_flat = jnp.ravel(loc)
  lb = jnp.asarray(0, dtype=np.float32)
  ub = jnp.asarray(100, dtype=np.float32)

  def body_fn(args):
      mu_i, loc_i = args
      print(lb, ub, mu_i, loc_i, _sum_over_range(mu_i, loc_i, lb, ub, 1))
      return _sum_over_range(mu_i, loc_i, lb, ub, 1)
  combined = jnp.stack([mu_flat, loc_flat], axis=-1)
  result = map(body_fn, combined)

  flat = jnp.stack(result)
  result = jnp.reshape(flat, mu.shape)
  return result

def _sum_over_range(mu, loc, start, stop, inc):
  fun = lambda x: entr(pmf(x, mu, loc))
  xs = jnp.arange(start, stop, inc)  # I feel like this may be an issue as the start should account for the loc offset but scipy doesn't account for it either, and adding it doesn't work.
  xs = jnp.floor(xs)
  vals = fun(xs)
  return jnp.sum(vals)
```
